streams/shared/src/main/scala/zio/stream/ZSummary.scala

```diff
@@ -0,0 +1,333 @@
+   * elements and signals when the sink should emit values. Whenever the `emitWhen`
+   * summary evalutes to `true`, the value from this summary will be emitted and the state
+   * for both summaries will be reset.
+   */
+  def toSink[I1 <: I](emitWhen: Summary[I1, Boolean]): ZSink[Any, Nothing, I1, I1, O] =
+    (self, emitWhen) match {
+      case (self: Summary.Mutable[I, O], emitWhen: Summary.Mutable[I1, Boolean]) =>
+        ZSink {
+          self.create.zip(emitWhen.create).toManaged_.map {
+            case (builder, emitBuilder) => {
+              case None =>
+                ZIO.effectSuspendTotal {
+                  val result = builder.extract()
+                  builder.reset()
+                  emitBuilder.reset()
+
+                  Sink.Push.emit(result, Chunk.empty)
+                }
+
+              case Some(is) =>
+                ZIO.effectSuspendTotal {
+                  var shouldEmit = false
+
+                  is.foreachWhile { i =>
+                    builder.step(i)
+                    emitBuilder.step(i)
+
+                    shouldEmit = emitBuilder.extract()
+
+                    !shouldEmit
+                  }
+
+                  val (_, leftover) = is.splitWhere { i =>
+                    builder.step(i)
+                    emitBuilder.step(i)
+
+                    shouldEmit = emitBuilder.extract()
+                    shouldEmit
+                  }
+
+                  if (!shouldEmit) Sink.Push.more
+                  else Sink.Push.emit(builder.extract(), leftover)
+                }
+            }
+          }
+
+        }
+    }
+
+  /**
+   * Composes two summaries by feeding incoming elements to both of them, and
+   * transforming their results with the provided function.
+   */
+  def zipWith[I1 <: I, O2, O3](that: Summary[I1, O2])(f: (O, O2) => O3): Summary[I1, O3] =
+    (self, that) match {
+      case (self: Summary.Mutable[I, O], that: Summary.Mutable[I1, O2]) =>
+        new Summary.Mutable[I1, O3] {
+          def create: zio.UIO[Summary.Builder[I1, O3]] =
+            self.create.zipWith(that.create) { (selfBuilder, thatBuilder) =>
+              new Summary.Builder[I1, O3] {
+                override def reset(): Unit = {
+                  selfBuilder.reset()
+                  thatBuilder.reset()
+                }
+
+                override def step(i: I1): Unit = {
+                  selfBuilder.step(i)
+                  thatBuilder.step(i)
+                }
+
+                override def extract(): O3 =
+                  f(selfBuilder.extract(), thatBuilder.extract())
+              }
+            }
+        }
+    }
+}
+
+object Summary {
+
+  /**
+   * A summary that produces a mutable builder that is used for
+   * summarizing the inputs.
+   */
+  abstract class Mutable[-I, +O] extends Summary[I, O] {
+
+    /**
+     * Creates a new mutable builder.
+     */
+    def create: UIO[Builder[I, O]]
```

Comment:
Maybe we should keep builder methods package private, and for users, expose a driver-like concept?

Reply:
Of course, internally, we'd use the "unsafe" methods, not the "pure" driver methodl